lib/gauge_random.cu

```diff
@@ -11,196 +11,183 @@
 
-      private:
-      unsigned int minThreads() const { return arg.threads; }
-      bool tuneGridDim() const { return false; } // Don't tune the grid dimensions.
+private:
+    unsigned int minThreads() const { return arg.threads; }
+    bool tuneGridDim() const { return false; } // Don't tune the grid dimensions.
 
-      public:
-      GaugeGauss(GaugeGaussArg<Gauge> &arg, GaugeField &gf)
-        : TunableVectorY(2), arg(arg), gf(gf){}
-      ~GaugeGauss () { }
+public:
+    GaugeGauss(Arg &arg, GaugeField &meta) : TunableVectorY(2), arg(arg), meta(meta) {}
+    ~GaugeGauss() {}
 
-      void apply(const cudaStream_t &stream){
-        if(gf.Location() == QUDA_CUDA_FIELD_LOCATION){
-          TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
-
-          computeGenGauss<Float><<<tp.grid,tp.block,tp.shared_bytes>>>(arg);
-	  qudaDeviceSynchronize();
-        } else {
-          errorQuda("Randomize GaugeFields on CPU not supported yet\n");
-        }
+    void apply(const cudaStream_t &stream)
+    {
+      if (meta.Location() == QUDA_CUDA_FIELD_LOCATION) {
+        TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+        computeGenGauss<Float><<<tp.grid, tp.block, tp.shared_bytes>>>(arg);
+      } else {
+        errorQuda("Randomize GaugeFields on CPU not supported yet\n");
       }
+    }
 
-      TuneKey tuneKey() const {
-        std::stringstream vol, aux;
-        vol << arg.X[0] << "x" << arg.X[1] << "x" << arg.X[2] << "x" << arg.X[3];
-	aux << "threads=" << arg.threads << ",prec="  << sizeof(Float);
-        return TuneKey(vol.str().c_str(), typeid(*this).name(), aux.str().c_str());
-      }
+    TuneKey tuneKey() const { return TuneKey(meta.VolString(), typeid(*this).name(), meta.AuxString()); }
+
+    long long flops() const { return 0; }
+    long long bytes() const { return meta.Bytes(); }
 
-      long long flops() const { return 0; }
-      long long bytes() const { return 0; } 
+    void preTune() { arg.rngstate.backup(); }
+    void postTune() { arg.rngstate.restore(); }
+  };
 
+  template <typename Float, QudaReconstructType recon, bool group>
+  void genGauss(GaugeField &U, RNG &rngstate, double beta)
+  {
+    GaugeGaussArg<Float, recon, group> arg(U, rngstate, beta);
+    GaugeGauss<Float, decltype(arg)> gaugeGauss(arg, U);
+    gaugeGauss.apply(0);
+  }
 
-      void preTune(){
-	  //gf.backup();
-	  arg.rngstate.backup();
+  template <typename Float> void gaugeGauss(GaugeField &U, RNG &rngstate, double beta)
+  {
+    if (U.LinkType() == QUDA_SU3_LINKS) { // generate Gaussian distributed SU(3) field
+      if (getVerbosity() >= QUDA_SUMMARIZE)
+        printfQuda("Creating Gaussian distrbuted gauge field with beta = %e\n", beta);
+      switch (U.Reconstruct()) {
+      case QUDA_RECONSTRUCT_NO: genGauss<Float, QUDA_RECONSTRUCT_NO, true>(U, rngstate, beta); break;
+      case QUDA_RECONSTRUCT_12: genGauss<Float, QUDA_RECONSTRUCT_12, true>(U, rngstate, beta); break;
```

Comment:
Let's add a `QUDA_RECONSTRUCT_13` and `QUDA_RECONSTRUCT_9` version of this, which would just alias to `QUDA_RECONSTRUCT_12` and `QUDA_RECONSTRUCT_8`, respectively? (Since we're explicitly requesting SU(3) fields, the extra phase would just be `theta = 0`.)

Reply:
We can't alias them to recon-12 and recon-8 since then `theta` would not be set to `0`.  We need to use their actual accessors and have the accessors compute `theta=0`.  Happy to make this change though if desired.

Reply:
May as well while we're here. I'll probably run into it head-first when I port some of these things to the staggered tests, too. 

Related---should we add templating over recon-9 and -13 for the internal plaquette computation routines while we're here? I don't think they've been instantiated.

Reply:
Ok, will do both of these.